File: patched-dex/src/main/java/com/samsung/android/shealthmonitor/ecg/manager/CalibrationState.java
package com.samsung.android.shealthmonitor.ecg.manager;

import android.content.Context;
import com.samsung.android.shealthmonitor.bp.R;
import com.samsung.android.shealthmonitor.bp.control.BpReCalibrationController;
import com.samsung.android.shealthmonitor.ecg.helper.EcgSharedPreferenceHelper;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorCompleteView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepCuffWearingVideoView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepCuffWearingView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepEmptyView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepMeasureMonitorView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepStartView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepView;
import com.samsung.android.shealthmonitor.bp.ui.view.SHealthMonitorStepWatchWearingVideoView;
import kotlin.jvm.internal.Intrinsics;

/* compiled from: CalibrationState.kt */
public enum CalibrationState {
    NONE {
		@Override
        public final int getCalibrationCount() {
            return 0;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return false;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return this;
        }

		@Override
        public final CalibrationState nextState() {
            return this;
        }

		@Override
        public final CalibrationState initState() {
            return this;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepEmptyView(context);
        }
    },
    FIRST_READY {
		@Override
        public final int getCalibrationCount() {
            return 1;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState nextState() {
            // return CalibrationState.FIRST_1_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.NONE;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepWatchWearingVideoView(context);
        }
    },

    FIRST_1_STEP {
		@Override
        public final int getCalibrationCount() {
            return 1;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.FIRST_READY;
        }

		@Override
        public final CalibrationState nextState() {
            // return CalibrationState.FIRST_2_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.FIRST_READY;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorCompleteView(context);
        }
		// @Override
        // public final SHealthMonitorStepView responsibleView(Context context) {
            // Intrinsics.checkParameterIsNotNull(context, "context");
            // return new SHealthMonitorStepCuffWearingVideoView(context, 0, 2, (DefaultConstructorMarker) null);
        // }
    },

    FIRST_2_STEP {
		@Override
        public final int getCalibrationCount() {
            return 1;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.FIRST_1_STEP;
        }

		@Override
        public final CalibrationState nextState() {
            // return CalibrationState.FIRST_DONE;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.FIRST_READY;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepMeasureMonitorView(context);
        }
    },

    FIRST_DONE {
		@Override
        public final int getCalibrationCount() {
            return 2;
        }

		@Override
        public final boolean hasMenu() {
            return true;
        }

		@Override
        public final boolean isPossibleBack() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState nextState() {
            // CalibrationState.SECOND_1_STEP.setPreviousStep(this);
            // return CalibrationState.SECOND_1_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.NONE;
        }

		@Override
        public final boolean save() {
            EcgSharedPreferenceHelper.setEcgCalibrationState(CalibrationState.SECOND_READY_STEP.name());
            EcgSharedPreferenceHelper.setEcgCurrentCalibrationStepTime(System.currentTimeMillis());
            BpReCalibrationController.initBpCalibrationNoti();
            return true;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorCompleteView(context);
        }
		// @Override
        // public final SHealthMonitorStepView responsibleView(Context context) {
            // Intrinsics.checkParameterIsNotNull(context, "context");
            // return new SHealthMonitorStepStartView(context, this);
        // }
    },

    SECOND_READY_STEP {
		@Override
        public final int getCalibrationCount() {
            return 2;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState nextState() {
            // CalibrationState.SECOND_1_STEP.setPreviousStep(this);
            // return CalibrationState.SECOND_1_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.NONE;
        }

		@Override
        public final boolean save() {
            EcgSharedPreferenceHelper.setEcgCalibrationState(name());
            return true;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepWatchWearingVideoView(context);
        }
    },

    SECOND_1_STEP {
		@Override
        public final int getCalibrationCount() {
            return 2;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState nextState() {
            // CalibrationState.SECOND_2_STEP.setPreviousStep(this);
            // return CalibrationState.SECOND_2_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.FIRST_DONE;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorCompleteView(context);
        }
		// @Override
        // public final SHealthMonitorStepView responsibleView(Context context) {
            // Intrinsics.checkParameterIsNotNull(context, "context");
            // if (previousState() == CalibrationState.SECOND_READY_STEP) {
                // return new SHealthMonitorStepCuffWearingVideoView(context, 0, 2, (DefaultConstructorMarker) null);
            // }
            // return new SHealthMonitorStepCuffWearingView(context, R.string.shealth_monitor_calibration_measurement_finish_get_ready_body);
        // }
    },

    SECOND_2_STEP {
		@Override
        public final int getCalibrationCount() {
            return 2;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState nextState() {
            // return CalibrationState.SECOND_DONE;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.SECOND_1_STEP;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepMeasureMonitorView(context);
        }
    },

    SECOND_DONE {
		@Override
        public final int getCalibrationCount() {
            return 3;
        }

		@Override
        public final boolean hasMenu() {
            return true;
        }

		@Override
        public final boolean isPossibleBack() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState nextState() {
            // CalibrationState.THIRD_1_STEP.setPreviousStep(this);
            // return CalibrationState.THIRD_1_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.NONE;
        }

		@Override
        public final boolean save() {
            EcgSharedPreferenceHelper.setEcgCalibrationState(CalibrationState.THIRD_READY_STEP.name());
            return true;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorCompleteView(context);
        }
		// @Override
        // public final SHealthMonitorStepView responsibleView(Context context) {
            // Intrinsics.checkParameterIsNotNull(context, "context");
            // return new SHealthMonitorStepStartView(context, this);
        // }
    },

    THIRD_READY_STEP {
		@Override
        public final int getCalibrationCount() {
            return 3;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState nextState() {
            // CalibrationState.THIRD_1_STEP.setPreviousStep(this);
            // return CalibrationState.THIRD_1_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.NONE;
        }

		@Override
        public final boolean save() {
            EcgSharedPreferenceHelper.setEcgCalibrationState(name());
            return true;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepWatchWearingVideoView(context);
        }
    },
    THIRD_1_STEP {
		@Override
        public final int getCalibrationCount() {
            return 3;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState nextState() {
            // CalibrationState.THIRD_2_STEP.setPreviousStep(this);
            // return CalibrationState.THIRD_2_STEP;
			return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.SECOND_DONE;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorCompleteView(context);
        }
		// @Override
        // public final SHealthMonitorStepView responsibleView(Context context) {
            // Intrinsics.checkParameterIsNotNull(context, "context");
            // if (previousState() == CalibrationState.THIRD_READY_STEP) {
                // return new SHealthMonitorStepCuffWearingVideoView(context, 0, 2, (DefaultConstructorMarker) null);
            // }
            // return new SHealthMonitorStepCuffWearingView(context, R.string.shealth_monitor_calibration_measurement_finish_get_ready_body);
        // }
    },
    THIRD_2_STEP {
		@Override
        public final int getCalibrationCount() {
            return 3;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return true;
        }

		@Override
        public final boolean save() {
            return false;
        }

		@Override
        public final CalibrationState nextState() {
            return CalibrationState.CALIBRATION_DONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.THIRD_1_STEP;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorStepMeasureMonitorView(context);
        }
    },

    CALIBRATION_DONE {
		@Override
        public final int getCalibrationCount() {
            return 4;
        }

		@Override
        public final boolean hasMenu() {
            return false;
        }

		@Override
        public final boolean isPossibleBack() {
            return false;
        }

		@Override
        public final CalibrationState previousState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState nextState() {
            return CalibrationState.NONE;
        }

		@Override
        public final CalibrationState initState() {
            return CalibrationState.NONE;
        }

		@Override
        public final boolean save() {
            EcgSharedPreferenceHelper.setEcgCalibrationState(name());
            EcgSharedPreferenceHelper.setEcgCurrentCalibrationStepTime(-1);
            BpReCalibrationController.removeBpCalibrationNoti();
            return true;
        }

		@Override
        public final SHealthMonitorStepView responsibleView(Context context) {
            Intrinsics.checkParameterIsNotNull(context, "context");
            return new SHealthMonitorCompleteView(context);
        }
    },
	;
    
    private CalibrationState mPreviousStep;

    public abstract int getCalibrationCount();

    public abstract boolean hasMenu();

    public abstract CalibrationState initState();

    public abstract boolean isPossibleBack();

    public abstract CalibrationState nextState();

    public abstract SHealthMonitorStepView responsibleView(Context context);

    public abstract boolean save();


    public CalibrationState getMPreviousStep() {
        return this.mPreviousStep;
    }

    public void setMPreviousStep(CalibrationState calibrationState) {
        this.mPreviousStep = calibrationState;
    }

    public void setPreviousStep(CalibrationState calibrationState) {
        Intrinsics.checkParameterIsNotNull(calibrationState, "state");
        setMPreviousStep(calibrationState);
    }

    public CalibrationState previousState() {
        if (getMPreviousStep() == null) {
            return NONE;
        }
        CalibrationState mPreviousStep2 = getMPreviousStep();
        if (mPreviousStep2 == null) {
            Intrinsics.throwNpe();
        }
        return mPreviousStep2;
    }
}

File: patched-dex/shm161_mymod_mod/unknown/com/samsung/android/shealthmonitor/bp/manager/CalibrationState.kt
package com.samsung.android.shealthmonitor.bp.manager

import android.content.Context
import com.samsung.android.shealthmonitor.bp.R
import com.samsung.android.shealthmonitor.bp.control.BpReCalibrationController
import com.samsung.android.shealthmonitor.bp.helper.BpSharedPreferenceHelper
import com.samsung.android.shealthmonitor.bp.ui.view.*

enum class CalibrationState {

    NONE {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            return NONE
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return false
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 0
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepEmptyView(context)
        }
    },
    FIRST_READY {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            return FIRST_1_STEP
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 1
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepWatchWearingVideoView(context)
        }
    },
    FIRST_1_STEP {
        override fun previousState(): CalibrationState {
            return FIRST_READY
        }
        override fun nextState(): CalibrationState {
            return FIRST_2_STEP
        }
        override fun initState(): CalibrationState {
            return FIRST_READY
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 1
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepCuffWearingVideoView(context)
        }
    },
    FIRST_2_STEP {
        override fun previousState(): CalibrationState {
            return FIRST_1_STEP
        }
        override fun nextState(): CalibrationState {
            return FIRST_DONE
        }
        override fun initState(): CalibrationState {
            return FIRST_READY
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 1
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepMeasureMonitorView(context)
        }
    },
    FIRST_DONE  {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            SECOND_1_STEP.setPreviousStep(FIRST_DONE)
            return SECOND_1_STEP
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return false
        }
        override fun hasMenu(): Boolean {
            return true
        }
        override fun getCalibrationCount() : Int {
            return 2
        }
        override fun save() : Boolean {
            BpSharedPreferenceHelper.setBpCalibrationState(SECOND_READY_STEP.name)
            BpSharedPreferenceHelper.setBpCurrentCalibrationStepTime(System.currentTimeMillis())
            BpReCalibrationController.initBpCalibrationNoti()
            return true
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            val view = SHealthMonitorStepStartView(context, FIRST_DONE)
            return view
        }
    },
    SECOND_READY_STEP {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            SECOND_1_STEP.setPreviousStep(SECOND_READY_STEP)
            return SECOND_1_STEP
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 2
        }
        override fun save() : Boolean {
            BpSharedPreferenceHelper.setBpCalibrationState(name)
            return true
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepWatchWearingVideoView(context)
        }
    },
    SECOND_1_STEP  {
        override fun nextState(): CalibrationState {
            SECOND_2_STEP.setPreviousStep(SECOND_1_STEP)
            return SECOND_2_STEP
        }
        override fun initState(): CalibrationState {
            return FIRST_DONE
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 2
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            if (previousState() == SECOND_READY_STEP) {
                return SHealthMonitorStepCuffWearingVideoView(context)
            } else {
                return SHealthMonitorStepCuffWearingView(context, R.string.shealth_monitor_calibration_measurement_finish_get_ready_body)
            }
        }
    },
    SECOND_2_STEP  {
        override fun nextState(): CalibrationState {
            return SECOND_DONE
        }
        override fun initState(): CalibrationState {
            return SECOND_1_STEP
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 2
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepMeasureMonitorView(context)
        }
    },
    SECOND_DONE {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            THIRD_1_STEP.setPreviousStep(SECOND_DONE)
            return THIRD_1_STEP
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return false
        }
        override fun hasMenu(): Boolean {
            return true
        }
        override fun getCalibrationCount() : Int {
            return 3
        }
        override fun save() : Boolean {
            BpSharedPreferenceHelper.setBpCalibrationState(THIRD_READY_STEP.name)
            return true
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            val view = SHealthMonitorStepStartView(context, SECOND_DONE)
            return view
        }
    },
    THIRD_READY_STEP {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            THIRD_1_STEP.setPreviousStep(THIRD_READY_STEP)
            return THIRD_1_STEP
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 3
        }
        override fun save() : Boolean {
            BpSharedPreferenceHelper.setBpCalibrationState(name)
            return true
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepWatchWearingVideoView(context)
        }
    },
    THIRD_1_STEP {
        override fun nextState(): CalibrationState {
            THIRD_2_STEP.setPreviousStep(THIRD_1_STEP)
            return THIRD_2_STEP
        }
        override fun initState(): CalibrationState {
            return SECOND_DONE
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 3
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            if (previousState() == THIRD_READY_STEP) {
                return SHealthMonitorStepCuffWearingVideoView(context)
            } else {
                return SHealthMonitorStepCuffWearingView(context, R.string.shealth_monitor_calibration_measurement_finish_get_ready_body)
            }
        }
    },
    THIRD_2_STEP {
        override fun nextState(): CalibrationState {
            return CALIBRATION_DONE
        }
        override fun initState(): CalibrationState {
            return THIRD_1_STEP
        }
        override fun isPossibleBack(): Boolean {
            return true
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 3
        }
        override fun save() : Boolean {
            return false
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorStepMeasureMonitorView(context)
        }
    },
    CALIBRATION_DONE {
        override fun previousState(): CalibrationState {
            return NONE
        }
        override fun nextState(): CalibrationState {
            return NONE
        }
        override fun initState(): CalibrationState {
            return NONE
        }
        override fun isPossibleBack(): Boolean {
            return false
        }
        override fun hasMenu(): Boolean {
            return false
        }
        override fun getCalibrationCount() : Int {
            return 4
        }
        override fun save() : Boolean {
            BpSharedPreferenceHelper.setBpCalibrationState(name)
            BpSharedPreferenceHelper.setBpCurrentCalibrationStepTime(-1)
            BpReCalibrationController.removeBpCalibrationNoti()
            return true
        }
        override fun responsibleView(context : Context): SHealthMonitorStepView {
            return SHealthMonitorCompleteView(context)
        }
    };

    open var mPreviousStep : CalibrationState? = null
    open fun setPreviousStep (state : CalibrationState) {
        mPreviousStep = state
    }

    open fun previousState() : CalibrationState {
        if (mPreviousStep == null) {
            return NONE
        } else {
            return mPreviousStep!!
        }
    }

    abstract fun nextState(): CalibrationState
    abstract fun initState(): CalibrationState
    abstract fun save() : Boolean
    abstract fun responsibleView(context : Context): SHealthMonitorStepView
    abstract fun isPossibleBack() : Boolean
    abstract fun hasMenu() : Boolean
    abstract fun getCalibrationCount() : Int

}